lib/segment/src/index/hnsw_index/hnsw.rs

```diff
@@ -317,61 +269,151 @@ impl<TGraphLinks: GraphLinks> HNSWIndex<TGraphLinks> {
+        params: Option<&SearchParams>,
+        is_stopped: &'a AtomicBool,
+    ) -> Box<dyn RawScorer + 'a> {
+        let quantization_enabled = Self::is_quantized_search(quantized_storage, params);
+        match quantized_storage {
+            Some(quantized_storage) if quantization_enabled => quantized_storage.raw_scorer(
+                vector.to_owned(),
+                id_tracker.deleted_point_bitslice(),
+                vector_storage.deleted_vector_bitslice(),
+                is_stopped,
+            ),
+            _ => new_stoppable_raw_scorer(
+                vector.to_owned(),
+                vector_storage,
+                id_tracker.deleted_point_bitslice(),
+                is_stopped,
+            ),
+        }
+    }
+
+    fn get_oversampled_top(
+        quantized_storage: Option<&QuantizedVectors>,
+        params: Option<&SearchParams>,
+        top: usize,
+    ) -> usize {
+        let quantization_enabled = Self::is_quantized_search(quantized_storage, params);
+
+        let oversampling_value = params
+            .and_then(|p| p.quantization)
+            .map(|q| q.oversampling)
+            .unwrap_or(default_quantization_oversampling_value());
+
+        match oversampling_value {
+            Some(oversampling) if quantization_enabled && oversampling > 1.0 => {
+                (oversampling * top as f64) as usize
+            }
+            _ => top,
         }
     }
+
+    fn postprocess_search_result(
+        &self,
+        search_result: Vec<ScoredPointOffset>,
+        vector: &QueryVector,
+        params: Option<&SearchParams>,
+        top: usize,
+        is_stopped: &AtomicBool,
+    ) -> Vec<ScoredPointOffset> {
+        let id_tracker = self.id_tracker.borrow();
+        let vector_storage = self.vector_storage.borrow();
+        let quantized_storage = vector_storage.quantized_storage();
+
+        let quantization_enabled = Self::is_quantized_search(quantized_storage, params);
+
+        let default_rescoring = quantized_storage
+            .map(|q| q.default_rescoring())
+            .unwrap_or(false);
+        let rescore = quantization_enabled
+            && params
+                .and_then(|p| p.quantization)
+                .and_then(|q| q.rescore)
+                .unwrap_or(default_rescoring);
+
+        let mut re_scored = if rescore {
+            let raw_scorer = new_stoppable_raw_scorer(
+                vector.to_owned(),
+                &vector_storage,
+                id_tracker.deleted_point_bitslice(),
+                is_stopped,
+            );
+
+            let mut ids_iterator = search_result.iter().map(|x| x.idx);
+            let mut re_scored = raw_scorer.score_points_unfiltered(&mut ids_iterator);
+
+            re_scored.sort_unstable();
+            re_scored.reverse();
+            re_scored.truncate(top);
+            re_scored
+        } else {
+            search_result
+        };
+        re_scored.truncate(top);
```

Comment:
We have two truncates here. We can probably remove one of them.

Reply:
agreed

Reply:
Great finding! Fixed